### PR TITLE
[fungible_assets] fix bugs, add tests, rename and change object api

### PR DESCRIPTION
**Pull request by lightmark**
### Description

1. rename wallet to store. rename FungibleAssetSupport to DeriveRefPod.
2. add tests to primary_store. Also fixed a bug there.
3. change create_user_derived_object to be a friend function w/o signer but use address instead.

### Test Plan
ut

### CODE REVIEW

File: aptos-move/move-examples/fungible_asset/sources/managed_fungible_asset.move [hunk 1]
```diff
@@ -26,7 +26,7 @@ module fungible_asset::managed_fungible_asset {
     /// Initialize metadata object and store the refs.
     fun init_module(admin: &signer) {
         let constructor_ref = &object::create_named_object(admin, ASSET_SYMBOL);
-        primary_wallet::create_primary_wallet_enabled_fungible_asset(
+        primary_store::create_primary_wallet_enabled_fungible_asset(
```
Comment (movekevin) [inline aptos-move/move-examples/fungible_asset/sources/managed_fungible_asset.move:29]: This looks broken. There's no create_primary_wallet_enabled_fungible_asset anymore

File: aptos-move/framework/aptos-framework/sources/primary_store.move [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    struct DeriveRefPod has key {
+        metadata_derive_ref: DeriveRef,
+    }
+
+    /// Creators of fungible assets can call this to enable support for creating primary (deterministic) stores for
+    /// their users.
+    public fun add_derived_ref_pod(
+        constructor_ref: &ConstructorRef,
+    ) {
+        let metadata_obj = &object::generate_signer(constructor_ref);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_store.move:19]: Why did you remove the call to fungible_asset::add_fungibility?

File: aptos-move/framework/aptos-framework/sources/object.move [hunk 3]
```diff
@@ -190,8 +192,7 @@ module aptos_framework::object {
 
     /// Create a new object whose address is derived based on the creator account address and another object.
     /// Derivde objects, similar to named objects, cannot be deleted.
-    public fun create_user_derived_object(creator: &signer, derive_ref: &DeriveRef): ConstructorRef {
-        let creator_address = signer::address_of(creator);
+    public(friend) fun create_user_derived_object(creator_address: address, derive_ref: &DeriveRef): ConstructorRef {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/object.move:195]: Why?

File: aptos-move/framework/aptos-framework/sources/primary_store.move [hunk 4]
```diff
@@ -0,0 +1,156 @@
+/// This defines the module for interacting with primary stores of accounts/objects, which have deterministic addresses
+module aptos_framework::primary_store {
+    use aptos_framework::fungible_asset::{Self, FungibleAsset, FungibleAssetStore};
+    use aptos_framework::object::{Self, Object, ConstructorRef, DeriveRef};
+
+    use std::signer;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Resource stored on the fungible asset metadata object to allow creating primary stores for it.
+    struct DeriveRefPod has key {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_store.move:10]: Why is DeriveRefPod a better name?

File: aptos-move/framework/aptos-framework/sources/primary_store.move [hunk 5]
```diff
@@ -0,0 +1,156 @@
+/// This defines the module for interacting with primary stores of accounts/objects, which have deterministic addresses
+module aptos_framework::primary_store {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_store.move:2]: Why is store > wallet?

File: aptos-move/framework/aptos-framework/sources/primary_store.move (hunk 2 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/primary_store.move:25]: change it back.
the assumption changes after I make the derived_object function friend.

File: aptos-move/framework/aptos-framework/sources/primary_store.move (hunk 4 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/primary_store.move:11]: Because it can be used for other purpose. If you agree that function is friend function, I agree to change it back to the original name. Otherwise, it is just a holder of derived ref.

File: aptos-move/framework/aptos-framework/sources/primary_store.move (hunk 5 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/primary_store.move:2]: from different ppl including our metapixel friends. I think store is better than container since they are used to xxx_store already.

File: aptos-move/move-examples/fungible_asset/sources/managed_fungible_asset.move (hunk 1 above)
Comment (lightmark) [reply aptos-move/move-examples/fungible_asset/sources/managed_fungible_asset.move:29]: my bad.

File: aptos-move/framework/aptos-framework/sources/object.move (hunk 3 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/object.move:195]: you don't need a signer now.
Another rationale is offering users to use this derive_ref arbitrarily would potentially make difference resources using the same object. I am not sure this is something we want.
cc @davidiw @CapCap

File: aptos-move/framework/aptos-framework/sources/primary_store.move (hunk 4 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/primary_store.move:11]: @gregnazario Any naming suggestion here? :P

File: aptos-move/framework/aptos-framework/sources/object.move [hunk 6]
```diff
@@ -471,6 +472,14 @@ module aptos_framework::object {
     }
 
     /// Accessors
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/object.move:474]: This should be a // and not a ///

File: aptos-move/framework/aptos-framework/sources/primary_store.move (hunk 4 above)
Comment (gregnazario) [reply aptos-move/framework/aptos-framework/sources/primary_store.move:11]: I mean I would make it PrimaryWalletCapability, though I think we overloaded that word now...

PrimaryWalletDeriveRef

Tells you what it is, and what it's used for.  Do you suspect it'll be used for anything else in the future?  Then

FungibleAssetDeriveRef

**Review by gregnazario — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 7]
```diff
@@ -53,14 +76,14 @@ module aptos_framework::fungible_asset {
...
-    struct FungibleAsset has key {
+    /// The store object that holds fungible assets of a specific type associated with an account.
+    struct FungibleStore has key {
         /// The address of the base metadata object.
         metadata: Object<Metadata>,
         /// The balance of the fungible metadata.
         balance: u64,
         /// Fungible Assets transferring is a common operation, this allows for freezing/unfreezing accounts.
-        allow_ungated_transfer: bool,
+        allow_ungated_balance_transfer: bool,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:86]: allow_owner_balance_transfer?

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 8]
```diff
@@ -21,26 +23,47 @@ module aptos_framework::fungible_asset {
...
+    /// Name of the fungible asset metadata is too long
+    const ENAME_TOO_LONG: u64 = 15;
+    /// Symbol of the fungible asset metadata is too long
+    const ESYMBOL_TOO_LONG: u64 = 16;
+
+    //
+    // Constants
+    //
+
+    const MAX_NAME_LENGTH: u64 = 32;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:51]: How did you come up with these numbers?
